file_writer.go

```diff
@@ -208,19 +210,34 @@ func (f *FileWriter) Close() error {
 		lastBlock = f.block.GetB()
 	}
 
-	completeReq := &hdfs.CompleteRequestProto{
-		Src:        proto.String(f.name),
-		ClientName: proto.String(f.client.namenode.ClientName()),
-		Last:       lastBlock,
-	}
-	completeResp := &hdfs.CompleteResponseProto{}
+	// Retry on Complete request returning false
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	backOff := 50 * time.Millisecond
+	for {
+		completeReq := &hdfs.CompleteRequestProto{
+			Src:        proto.String(f.name),
+			ClientName: proto.String(f.client.namenode.ClientName()),
+			Last:       lastBlock,
+		}
+		completeResp := &hdfs.CompleteResponseProto{}
 
-	err := f.client.namenode.Execute("complete", completeReq, completeResp)
-	if err != nil {
-		return &os.PathError{"create", f.name, err}
+		err := f.client.namenode.Execute("complete", completeReq, completeResp)
+		if err != nil {
+			return &os.PathError{"create", f.name, err}
+		}
+		// Return no error if fileWriter is successfully closed.
+		if closed := completeResp.GetResult(); closed {
+			return nil
+		}
+		// If fileWriter is not yet closed, wait for backOff and retry or timeout.
+		select {
+		case <-time.After(backOff):
+			backOff *= 2
```

Comment:
I think you should cap the back off to 2 seconds.

Reply:
Done